plugins/genesys/fu-genesys-hubhid-device.c

```diff
@@ -0,0 +1,351 @@
+			    sizeof(FuGenesysUsbSetup),
+			    error))
+		return FALSE;
+
+	/* send request report */
+	if (!fu_hid_device_set_report(hid_device,
+				      GENESYS_HUBHID_REPORT_ID,
+				      buf_report->data,
+				      buf_report->len,
+				      GENESYS_HUBHID_REPORT_TIMEOUT,
+				      GENESYS_HUBHID_REPORT_FLAGS,
+				      error))
+		return FALSE;
+
+	if (setup->length == 0) {
+		g_warning("read zero-length hid report");
+		return TRUE;
+	}
+
+	/* receive report */
+	chunks = fu_chunk_array_mutable_new(data, setup->length, 0, 0x0, buf_report->len);
+	if (progress != NULL) {
+		fu_progress_set_id(progress, G_STRLOC);
+		fu_progress_set_steps(progress, chunks->len);
+	}
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+
+		memset(buf_report->data, 0, buf_report->len);
+		if (!fu_hid_device_get_report(hid_device,
+					      GENESYS_HUBHID_REPORT_ID,
+					      buf_report->data,
+					      buf_report->len,
+					      GENESYS_HUBHID_REPORT_TIMEOUT,
+					      GENESYS_HUBHID_REPORT_FLAGS |
+						  FU_HID_DEVICE_FLAG_RETRY_FAILURE,
+					      error)) {
+			g_prefix_error(error,
+				       "error getting report at 0x%04x: ",
+				       fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		if (!fu_memcpy_safe(fu_chunk_get_data_out(chk),
+				    fu_chunk_get_data_sz(chk),
+				    0, /* dst */
+				    buf_report->data,
+				    buf_report->len,
+				    0x0, /* src */
+				    fu_chunk_get_data_sz(chk),
+				    error)) {
+			g_prefix_error(error,
+				       "error getting report data at 0x%04x: ",
+				       fu_chunk_get_address(chk));
+			return FALSE;
+		}
+		if (progress != NULL)
+			fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_genesys_hubhid_device_can_pack_report(FuGenesysHubhidDevice *self, guint16 data_length)
+{
+	if (!self->support_report_pack)
+		return FALSE;
+
+	if (data_length > self->max_report_pack_data_length)
+		return FALSE;
+
+	return TRUE;
+}
+
+gboolean
+fu_genesys_hubhid_device_command_write(FuGenesysHubhidDevice *self,
+				       FuGenesysUsbSetup *setup,
+				       const guint8 *data,
+				       FuProgress *progress,
+				       GError **error)
+{
+	FuHidDevice *hid_device = FU_HID_DEVICE(self);
+	gboolean pack_report = FALSE;
+	g_autoptr(GByteArray) buf_report = g_byte_array_new();
+
+	fu_byte_array_set_size(buf_report, self->report_length, 0);
+
+	/* set request data */
+	if (!fu_memcpy_safe(buf_report->data,
+			    buf_report->len,
+			    0, /* dst */
+			    (const guint8 *)setup,
+			    sizeof(FuGenesysUsbSetup),
+			    0x0, /* src */
+			    sizeof(FuGenesysUsbSetup),
+			    error))
+		return FALSE;
+
+	/* pack report if it can */
+	pack_report = fu_genesys_hubhid_device_can_pack_report(self, setup->length);
+	if (pack_report) {
+		if (setup->length > 0 && !fu_memcpy_safe(buf_report->data,
+							 buf_report->len,
+							 sizeof(FuGenesysUsbSetup), /* dst */
+							 data,
+							 setup->length,
```

Comment:
is data always of length  setup->length? I'd feel a lot more comfortable with a parameter of `gsize datasz` to this function.

Reply:
> is data always of length setup->length?

data may larger than setup->length.

>  I'd feel a lot more comfortable with a parameter of gsize datasz to this function.

TL;DR:  I will check all vendor command caller to meet the goal of your wish.

Because this function is called from `fu_genesys_usbhub_device_ctrl_transfer`, and all vendor command calls `fu_genesys_usbhub_device_ctrl_transfer`. I think the vendor command caller has responsibility to maintain the `data` and `datasz`. 

I can add a comment for these two functions to indicate user has the responsibility of `data` and `setup->length`.

Reply:
I changed the `fu_genesys_usbhub_device_ctrl_transfer` interface to fulfill `g_usb_device_control_transfer`, and revert the caller usage. Can you check that is well?